pkg/registry/directoryGraphLoader.go

```diff
@@ -0,0 +1,223 @@
+func (g *DirGraphLoader) loadReplaces() {
+	sort.Sort(g.SortedCSVs)
+
+	for csvName, replace := range g.CsvNameAndReplaceMap {
+		if replace.skipRange != nil {
+			for _, v := range g.SortedCSVs {
+				if replace.skipRange(v.version) {
+					g.CsvNameAndReplaceMap[csvName] = csvReplaces{
+						replaces:  append(g.CsvNameAndReplaceMap[csvName].replaces, v.name),
+						skipRange: g.CsvNameAndReplaceMap[csvName].skipRange,
+					}
+				}
+			}
+		}
+	}
+}
+
+// Generate returns Package graph by parsing through package directory assuming all bundles in the package exist.
+func (g *DirGraphLoader) Generate() (*Package, error) {
+	pkg, err := g.parsePackageYAMLFile()
+	if err != nil {
+		return nil, fmt.Errorf("error parsing package.yaml file in the package root directory, %v", err)
+	}
+
+	for chName, ch := range pkg.Channels {
+		pkg.Channels[chName] = Channel{
+			Head:  ch.Head,
+			Nodes: *g.getChannelNodes(ch.Head.CsvName),
+		}
+	}
+	return pkg, nil
+}
+
+// getChannelNodes follows the head of the channel csv through all replaces to fill the nodes until the tail of the
+// channel which will not replace anything.
+func (g *DirGraphLoader) getChannelNodes(channelHeadCsv string) *map[BundleKey]map[BundleKey]struct{} {
+	nodes := make(map[BundleKey]map[BundleKey]struct{})
+	remainingCSVsInChannel := make(map[BundleKey]struct{})
+
+	for _, csv := range g.CsvNameAndReplaceMap[channelHeadCsv].replaces {
+		remainingCSVsInChannel[BundleKey{CsvName: csv}] = struct{}{}
+	}
+
+	// Iterate through remainingCSVsInChannel and add replaces of each encountered CSVs if not already in nodes.
+	// Loop only exit after all remaining csvs are visited/deleted.
+	for len(remainingCSVsInChannel) > 0 {
+		for bk, _ := range remainingCSVsInChannel {
+			if _, ok := nodes[BundleKey{CsvName: channelHeadCsv}]; !ok {
+				nodes[BundleKey{CsvName: channelHeadCsv}] = func() map[BundleKey]struct{} {
+					subNode := make(map[BundleKey]struct{})
+					for _, csv := range g.CsvNameAndReplaceMap[channelHeadCsv].replaces {
+						subNode[BundleKey{CsvName: csv}] = struct{}{}
+					}
+					return subNode
+				}()
+				for _, csv := range g.CsvNameAndReplaceMap[channelHeadCsv].replaces {
+					if _, ok := nodes[BundleKey{CsvName: csv}]; !ok {
+						remainingCSVsInChannel[BundleKey{CsvName: csv}] = struct{}{}
+					}
+				}
+			}
+			delete(remainingCSVsInChannel, bk)
+		}
+	}
+	return &nodes
+}
+
+// parsePackageYAMLFile parses the *.package.yaml file and fills the information in Package including name,
+// defaultchannel, and head of all Channels. It returns parsing error if any.
+func (g *DirGraphLoader) parsePackageYAMLFile() (*Package, error) {
+	files, err := ioutil.ReadDir(g.PackageDir)
+	if err != nil {
+		return nil, fmt.Errorf("error reading bundle parent directory, %v", err)
+	}
+
+	ymlFile, err := func() (string, error) {
+		for _, f := range files {
+			if matched, err := path.Match("*.package.yaml", f.Name()); err != nil {
```

Comment:
same comment as the parser for the csv, i think we can serialize this rather than rely on a naming convention